glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -1,73 +1,265 @@
+const END_OF_SCAN: u16 = NUM_OF_SLOTS;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE as usize];
+
+/// Holds configuration for a cluster scan operation.
+///
+/// # Fields
+/// - `scan_state_cursor`: Internal state tracking scan progress
+/// - `match_pattern`: Optional pattern to filter keys
+/// - `count`: Optional limit on number of keys returned per iteration
+/// - `object_type`: Optional filter for specific data types
+/// - `allow_non_covered_slots`: Whether to continue if some slots are uncovered
+///
+/// See examples below for usage with the builder pattern.
+/// # Examples
+///
+/// Create a new `ClusterScanArgs` instance using the builder pattern:
+///
+/// ```rust,no_run
+/// use redis::ClusterScanArgs;
+/// use redis::ObjectType;
+///
+/// // Create basic scan args with defaults
+/// let basic_scan = ClusterScanArgs::builder().build();
+///
+/// // Create scan args with custom options
+/// let custom_scan = ClusterScanArgs::builder()
+///     .with_match_pattern("user:*")      // Match keys starting with "user:"
+///     .with_count(100)                   // Return 100 keys per iteration
+///     .with_object_type(ObjectType::Hash) // Only scan hash objects
+///     .allow_non_covered_slots(true)     // Continue scanning even if some slots aren't covered
+///     .build();
+///
+/// // The builder can be used to create multiple configurations
+/// let another_scan = ClusterScanArgs::builder()
+///     .with_match_pattern("session:*")
+///     .with_object_type(ObjectType::String)
+///     .build();
+/// ```
 
-const BITS_PER_U64: usize = u64::BITS as usize;
-const NUM_OF_SLOTS: usize = SLOT_SIZE as usize;
-const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
-const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
-type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+#[derive(Clone, Default)]
+pub struct ClusterScanArgs {
+    /// Reference-counted scan state cursor, managed internally.
+    pub scan_state_cursor: ScanStateRC,
 
-#[derive(Clone)]
-pub(crate) struct ClusterScanArgs {
-    pub(crate) scan_state_cursor: ScanStateRC,
+    /// Optional pattern to match keys during the scan.
+    pub match_pattern: Option<Vec<u8>>,
+
+    /// Optional number of keys to return per scan.
+    pub count: Option<u32>,
+
+    /// Optional filter to include only keys of a specific data type.
+    pub object_type: Option<ObjectType>,
+
+    /// Flag indicating whether to allow scanning when there are slots not covered by the cluster.
+    pub allow_non_covered_slots: bool,
+}
+
+impl ClusterScanArgs {
+    /// Creates a new [`ClusterScanArgsBuilder`] instance.
+    ///
+    /// # Returns
+    ///
+    /// A [`ClusterScanArgsBuilder`] instance for configuring cluster scan arguments.
+    pub fn builder() -> ClusterScanArgsBuilder {
+        ClusterScanArgsBuilder::default()
+    }
+    pub(crate) fn set_scan_state_cursor(&mut self, scan_state_cursor: ScanStateRC) {
+        self.scan_state_cursor = scan_state_cursor;
```

Comment:
https://github.com/valkey-io/valkey-glide/pull/2815#discussion_r1888456523

nit: with_cursor_state

Answer:

Its different than the other, since its done internally and by user, its not part of the builder per se. What do you think will fit?
and it a cursor of the scan state, not a state of the cursor.